staging/src/k8s.io/apiserver/pkg/server/filters/shutdown_annotations.go

```diff
@@ -0,0 +1,63 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+type shutdownEvent interface {
+	// Name returns the name of the signal, useful for logging.
+	Name() string
+
+	// Signaled returns a channel that is closed when the underlying event
+	// has been signaled. Successive calls to Signaled return the same value.
+	Signaled() <-chan struct{}
+
+	// SignaledAt returns the time when the event was signaled. If it is
+	// called before the event is signaled zero is returned.
+	SignaledAt() time.Time
+}
+
+func WithShutdownAnnotation(handler http.Handler, initiated shutdownEvent, delayDuration time.Duration) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		select {
+		case <-initiated.Signaled():
+		default:
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		requestor, exists := request.UserFrom(req.Context())
+		if !exists {
+			responsewriters.InternalError(w, req, errors.New("no user found for request"))
+			return
+		}
+		if requestor.GetName() == user.APIServerUser {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		late := fmt.Sprintf("since=%s elapsed=%s threshold=%s", initiated.Name(), time.Since(initiated.SignaledAt()).String(), delayDuration)
```

Comment:
why does delayDuration matter here?

Reply:
this is the `shutdown-delay-duration` server run option, if we want to know how late compared to the grace period it will be helpful. The audit mining tool does not need to fetch the configuration value from a different source.